File: src/Skies.Framework.Abstractions/OrderedLifecycle.cs
using System;
using System.Collections.Generic;

namespace Skies.Framework.Abstractions;

/// <summary>
/// The mechanics of a step-ordered lifecycle. An entity that walks an onboarding/approval sequence declares
/// its states as an enum <em>in step order</em> and stores the furthest one reached; this helper is the two
/// operations that order implies, so each entity stops re-deriving them:
///
/// <list type="bullet">
/// <item><description><see cref="Reached{TState}"/> — has the cursor passed (or landed on) a step? The gate
/// for "may this step's data be (re)written" — a step not yet reached is rejected, so steps can't be
/// skipped, while a completed one can be revisited.</description></item>
/// <item><description><see cref="Advance{TState}"/> — move forward exactly one step, and only from the step
/// being finished, so completing the current step progresses while re-editing an earlier one never
/// regresses the cursor.</description></item>
/// </list>
///
/// The comparison relies on declaration order (the enum's underlying values), which is the same contract the
/// pilot's Host/Traveler lifecycles depended on when this logic was duplicated byte-for-byte across them.
/// </summary>
/// <example>
/// <code>
/// // On the entity, the cursor is `LifecycleState`:
/// public bool StepReached(HostState step) => OrderedLifecycle.Reached(LifecycleState, step);
/// private void AdvancePast(HostState step, HostState next) =>
///     LifecycleState = OrderedLifecycle.Advance(LifecycleState, step, next);
/// </code>
/// </example>
public static class OrderedLifecycle
{
    /// <summary>True when <paramref name="current"/> is at or past <paramref name="step"/> in declaration
    /// order — so the step has been reached and its data may be written or re-edited. A step ahead of the
    /// cursor returns false (no skipping).</summary>
    public static bool Reached<TState>(TState current, TState step)
        where TState : struct, Enum =>
        Comparer<TState>.Default.Compare(current, step) >= 0;

    /// <summary>Advance the cursor to <paramref name="next"/> only when it is exactly at <paramref name="step"/>
    /// — finishing the current step moves forward. From any other position (an earlier step being re-edited, or
    /// a later one) it returns <paramref name="current"/> unchanged, so the cursor never regresses or jumps.</summary>
    public static TState Advance<TState>(TState current, TState step, TState next)
        where TState : struct, Enum =>
        EqualityComparer<TState>.Default.Equals(current, step) ? next : current;
}

File: node-sdk/packages/core/src/ordered-lifecycle.ts
/** Operations for a lifecycle whose states have one explicit declaration order. */
export interface OrderedLifecycle<TState> {
  /** True when `current` is at or past `step`, allowing that step to be written or revisited. */
  reached(current: TState, step: TState): boolean;
  /**
   * Returns `next` only when the cursor is exactly at the step being completed and `next` is its declared
   * immediate successor. Earlier-step edits, regressions, and skipped states leave the cursor unchanged.
   */
  advance(current: TState, step: TState, next: TState): TState;
}

/**
 * Defines order-dependent lifecycle mechanics from an explicit tuple of states. Explicit order keeps
 * string-literal states usable without decorators, numeric enums, or hidden discovery.
 */
export function orderedLifecycle<const TState>(states: readonly TState[]): OrderedLifecycle<TState> {
  if (states.length === 0) throw new TypeError("An ordered lifecycle requires at least one state");

  const positions = new Map<TState, number>();
  states.forEach((state, index) => {
    if (positions.has(state)) throw new TypeError("An ordered lifecycle cannot contain duplicate states");
    positions.set(state, index);
  });

  const positionOf = (state: TState): number => {
    const position = positions.get(state);
    if (position === undefined) throw new TypeError("State does not belong to this ordered lifecycle");
    return position;
  };

  return {
    reached: (current, step) => positionOf(current) >= positionOf(step),
    advance: (current, step, next) => {
      const currentPosition = positionOf(current);
      const stepPosition = positionOf(step);
      const nextPosition = positionOf(next);
      const isExactNextStep = nextPosition === stepPosition + 1;
      return currentPosition === stepPosition && isExactNextStep ? next : current;
    },
  };
}
